### PR TITLE
renameAll and renamePath

### PR DESCRIPTION
**Pull request by flintinatux**
dilbert rename

### CODE REVIEW

File: src/rename.js [hunk 1]
```diff
@@ -2,10 +2,12 @@ const assoc    = require('ramda/src/assoc')
...
 const curry    = require('ramda/src/curry')
 const dissoc   = require('ramda/src/dissoc')
+const hasIn    = require('ramda/src/hasIn')
 const prop     = require('ramda/src/prop')
+const when     = require('ramda/src/when')
 
 // rename :: String -> String -> { k: v } -> { k: v }
-const rename = (from, to, obj) =>
-  converge(assoc(to), [ prop(from), dissoc(from) ])(obj)
+const rename = (frum, to, obj) =>
```
Comment (pklingem) [inline src/rename.js]: ok, I'll bite, why `frum`?

Comment (pklingem) [reply src/rename.js]: oh, because `from` is a reserved word in es6.

File: src/renameAll.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+const apply   = require('ramda/src/apply')
+const compose = require('ramda/src/compose')
+const curry   = require('ramda/src/curry')
+const map     = require('ramda/src/map')
+const toPairs = require('ramda/src/toPairs')
+
+const rename = require('./rename')
+
+// renameAll :: { k: v } -> { k: v } -> { k: v }
+const renameAll = (renames, obj) =>
```
Comment (pklingem) [inline src/renameAll.js]: am I crazy or could this just be:

```javascript
const renameAll =
  compose(apply(compose), map(apply(rename)), toPairs)

module.exports = curry(renameAll)
```

Comment (mgreystone) [reply src/renameAll.js]: Would need crocks.curry for that to work.

File: src/renameAll.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  compose(apply(compose), map(apply(rename)), toPairs)(renames)(obj)
```
Comment (mgreystone) [inline src/renameAll.js]: `apply(compose)` will error if `obj` only has <= 1 value, yes?

File: src/renameAll.js (hunk 2 above)
Comment (flintinatux) [reply src/renameAll.js]: Only if we pulled in `crocks#curry`.  Ramda's wouldn't do it right, and we haven't made `crocks` a dep of this library yet.

File: src/rename.js (hunk 1 above)
Comment (flintinatux) [reply src/rename.js]: yep.  even my highlighter colors it as a reserved word.  plus, `frum` makes your face look poutier when you say it.

**Comment by flintinatux** [comment]
@pklingem, by recursive, do you mean this?
```js
renameAll({ course: { id: 'courseId' } })
```
We can do that, sure, if you think it's of value.  I can be constructed with `evolve` if we don't have that:
```js
evolve({ course: renameAll({ id: 'courseId' }) })
```
Just say the word and I can make it happen.

File: src/renameAll.js (hunk 3 above)
Comment (flintinatux) [reply src/renameAll.js]:
```
> compose(Function.prototype)
[Function]
> compose()
Error: compose requires at least one argument
```
So it looks like `renameAll({})` wouldn't work, sure.  But you'd get such a nice descriptive error (see above) to say what you did wrong.

**Comment by pklingem** [comment]
I think it's worth doing, because someone will want it eventually.

File: src/renameAll.js (hunk 3 above)
Comment (pklingem) [reply src/renameAll.js]: you should be able to pass `{}` without error, imagine if that object is dynamically constructed, you wouldn't want it to error in that case.

Comment (flintinatux) [reply src/renameAll.js]: Sorry, I should have added the :trollface:.

Comment (pklingem) [reply src/renameAll.js]: lol, you fooled me

File: API.md [hunk 4]
```diff
@@ -422,12 +423,28 @@ rename :: String -> String -> { k: v } -> { k: v }
...
+```haskell
+renameAll :: { k: v } -> { k: v } -> { k: v }
+```
+
+Rename multiple properties on an object to have different keys.
+
+```js
+const orig = { first_name: 'Miles', last_name: 'Callisto' }
+renameAll({ first_name: 'firstName', last_name: 'lastName' }, orig)
+//=> { firstName: 'Miles', lastName: 'Callisto' }
```
Comment (rpearce) [inline API.md]: Maybe I'm missing something, but this seems backwards to me.
```js
const deIdentify =
  renameAll({ first_name: 'firstName', last_name: 'lastName' })

deIdentify({ first_name: 'Miles', last_name: 'Callisto' })
```
says to me that the result should be
```js
//=> { firstName: 'firstName', lastName: 'lastName' }
```
I believe the first argument should be what the outcome should be, then the last one is the `orig`inal data.

File: src/rename.js [hunk 5]
```diff
@@ -2,10 +2,12 @@ const assoc    = require('ramda/src/assoc')
...
+  when(hasIn(frum), converge(assoc(to), [ prop(frum), dissoc(frum) ]))(obj)
```
Comment (rpearce) [inline src/rename.js]: Do we need/want to have `obj` explicitly being passed in on the end?
```js
// rename :: String -> String -> { k: v } -> { k: v }
const rename = (frum, to) =>
  when(hasIn(frum), converge(assoc(to), [ prop(frum), dissoc(frum) ]))
```

**Comment by rpearce** [comment]
@pklingem @flintinatux if `renameAll` were to be able to dig into an object, wouldn't we want `rename` to do that, as well?

File: src/rename.js (hunk 5 above)
Comment (flintinatux) [reply src/rename.js]: If we pulled in `crocks.curry`, I'd be ok with this.  What I don't want to match is Ramda's inconsistency in currying: some of their functions curry the final argument, and some (like `juxt` and `converge`) do not.  I would like to always curry the final argument like all the others, and to do so we can either use `R.curry` and pass the `obj` explicitly, or use `crocks.curry` and let it do the magic.

**Comment by flintinatux** [comment]
@rpearce, when you say:
> wouldn't we want `rename` to do that, as well?

Perhaps you're thinking of another new function, `renamePath`.  Similar to `assocPath`, it could be: 
```haskell
renamePath :: [String] -> String -> { k: v }
```
`rename` is like `assoc`, it takes only a string as the first arg, and I think it should stay that way.  But we could add a matching `renamePath` if you like it'd be worthwhile.

**Comment by rpearce** [comment]
@flintinatux yeah `renamePath` okay :+1:

File: src/rename.js (hunk 5 above)
Comment (rpearce) [reply src/rename.js]: Okay maybe some day. And crocks' `curry` isn't magic! It's gorgeous. I break down each line in a blog post here:
https://robertwpearce.com/blog/ramda-chops-function-currying.html

**Comment by flintinatux** [comment]
@rpearce,
I enjoy making useful tools, so no worries.  But just to be double sure I understand, do you think:
```js
renamePath(['media', 'embed', 'type'], 'embedType')
```
would be clearer or more helpful enough in some cases not to use:
```js
renameAll({ media: { embed: { type: 'embedType' } } })
```

File: API.md (hunk 4 above)
Comment (rpearce) [reply API.md]: I'm so sorry; I'm still a little confused. The pseudo-type looks like this currently, right?
```haskell
OriginalObject -> RenameObject -> NewObject
```
whereas I'm asking if it should be:
```haskell
RenameObject -> OriginalObject  -> NewObject
```
So when you mention `{ frum: to }` above, I'm not quite sure what we're talking about, unfortunately. I'm also not as deep into this project, so I may have no clue what I'm talking about.

With what's currently here, if I want to compose, I'll need to do this?

```js
const deIdentify =
  renameAll(__, { first_name: 'firstName', last_name: 'lastName' })

const doStuff =
  compose(deIdentity, somethingElse)

doStuff({ first_name: 'Miles', last_name: 'Callisto', email: 'mc@mc.org' })
//=> { firstName: 'firstName', lastName: 'lastName', email: 'mc@mc.org' }
```

Comment (flintinatux) [reply API.md]: Yeah, now I'm confused.  `renameAll` already takes the renames first and the original second:
https://github.com/articulate/funky/blob/a02064529ca637f5289b78643e3750fccd13e959/src/renameAll.js#L4

Remember, we're renaming **keys**, not **values**.  Other existing functions like `R.merge` work on the values.

Comment (rpearce) [reply API.md]: ffs. what is wrong with me

Comment (rpearce) [reply API.md]: ship it

**Comment by rpearce** [comment]
@flintinatux missed this notification, and I'm now here to see this through.

In short, yes.

This looks great, and I am so sorry about my confusion last week with regard to this. I was moving too fast.

**Comment by flintinatux** [comment]
squirrel

**Pull request merged**